sklearn/linear_model/_glm/glm.py

```diff
@@ -20,13 +24,746 @@
+    decomposition of X':
+
+        X' = QR with Q'Q = identity(k), k = min(n_samples, n_features)
+
+    This is the same as an LQ decomposition of X. We introduce the new variable t as,
+    see [1]:
+
+        (coef, intercept) = (Q @ t, intercept)
+
+    By using X @ coef = R' @ t and ||coef||_2 = ||t||_2, we can just replace X
+    by R', solve for t instead of coef, and finally get coef = Q @ t.
+    Note that t has less elements than coef if n_features > n_samples:
+        len(t) = k = min(n_samples, n_features) <= n_features = len(coef).
+
+    [1] Hastie, T.J., & Tibshirani, R. (2003). Expression Arrays and the p n Problem.
+    https://web.stanford.edu/~hastie/Papers/pgtn.pdf
+    """
+
+    def setup(self, X, y, sample_weight):
+        n_samples, n_features = X.shape
+        # TODO: setting pivoting=True could improve stability
+        # QR of X'
+        self.Q, self.R = scipy.linalg.qr(X.T, mode="economic", pivoting=False)
+        # use k = min(n_features, n_samples) instead of n_features
+        k = self.R.T.shape[1]
+        n_dof = k
+        if self.linear_loss.fit_intercept:
+            n_dof += 1
+        # store original coef
+        self.coef_original = self.coef
+        # set self.coef = t (coef_original = Q @ t)
+        self.coef = np.zeros_like(self.coef, shape=n_dof)
+        if np.sum(np.abs(self.coef_original)) > 0:
+            self.coef[:k] = self.Q.T @ self.coef_original[:n_features]
+        self.gradient = np.empty_like(self.coef)
+        self.hessian = np.empty_like(self.coef, shape=(n_dof, n_dof))
+
+        super().setup(X=self.R.T, y=y, sample_weight=sample_weight)
+
+    def update_gradient_hessian(self, X, y, sample_weight):
+        # Use R' instead of X
+        self.linear_loss.gradient_hessian(
+            coef=self.coef,
+            X=self.R.T,
+            y=y,
+            sample_weight=sample_weight,
+            l2_reg_strength=self.l2_reg_strength,
+            n_threads=self.n_threads,
+            gradient_out=self.gradient,
+            hessian_out=self.hessian,
+            raw_prediction=self.raw_prediction,  # this was updated in line_search
+        )
+
+    def line_search(self, X, y, sample_weight):
+        # Use R' instead of X
+        super().line_search(X=self.R.T, y=y, sample_weight=sample_weight)
+
+    def check_convergence(self, X, y, sample_weight):
+        # Use R' instead of X
+        super().check_convergence(X=self.R.T, y=y, sample_weight=sample_weight)
+
+    def finalize(self, X, y, sample_weight):
+        n_features = X.shape[1]
+        w, intercept = self.linear_loss.weight_intercept(self.coef)
+        self.coef_original[:n_features] = self.Q @ w
+        if self.linear_loss.fit_intercept:
+            self.coef_original[-1] = intercept
+        self.coef = self.coef_original
+
+
+class LSMRNewtonSolver(NewtonSolver):
+    """LSMR based Newton solver.
+
+    The inner solver uses LSMR after the Newton update is cast into the IRLS
```

Comment:
Could you please expand those acronyms and point to a standard reference for LSMR?

As far as I understand:

- LSMR stands for Least Squares with MINRES (as an alternative to Least Squares with QR) and is introduced in https://web.stanford.edu/group/SOL/software/lsmr/LSMR-SISC-2011.pdf
- IRLS: Iteratively Reweighted Least Squares